core/lens/lens_phrase.ml

```diff
@@ -0,0 +1,235 @@
+
+let tuple v = TupleLit v
+
+let tuple_singleton v = tuple [v]
+
+let name_of_var expr =
+  match expr.node with
+  | `Var n -> n
+  | _ -> failwith "Expected var."
+
+let of_phrase p =
+  let rec f p =
+    match p.node with
+    | `Constant c -> Constant c
+    | `Var v -> Var v
+    | `UnaryAppl ((_, op), phrase) -> UnaryAppl (Unary.from_links op, f phrase)
+    | `InfixAppl ((_, op), phrase1, phrase2) -> InfixAppl (Binary.from_links op, f phrase1, f phrase2)
+    | `TupleLit l -> TupleLit (List.map f l)
+    | `FnAppl (fn, arg) ->
+      begin
+        match name_of_var fn with
+        | "not" -> UnaryAppl ((Unary.Name "!"), f (List.hd arg))
+        | _ -> failwith "Unsupported function"
+      end
+    | _ -> failwith "Unknown phrasenode for lens_phrase to phrase."
+  in
+  f p
+
+let rec traverse (expr : lens_phrase) dosth =
+  let fn expr' = traverse expr' dosth in
+  let expr = match expr with
+    | Constant _ -> expr
+    | Var _ -> expr
+    | UnaryAppl (a, arg) ->
+      let arg = fn arg in
+      UnaryAppl (a, arg)
+    | InfixAppl (a, a1, a2) ->
+      let a1 = fn a1 in
+      let a2 = fn a2 in
+      InfixAppl (a, a1, a2)
+    | TupleLit (x :: []) ->
+      let x = fn x in
+      TupleLit ([x])
+    | Case (phr, cases, otherwise) ->
+      let phr = OptionUtils.opt_map fn phr in
+      let cases = List.map (fun (inp, lst) -> fn inp, fn lst) cases in
+      let otherwise = fn otherwise in
+      Case (phr, cases, otherwise)
+    | _ -> failwith "Unknown operation" in
+  dosth expr
+
+let get_vars expr =
+  let cols = ref Alias.Set.empty in
+  let _ = traverse expr (fun expr ->
+      match expr with
+      | Var n -> cols := Alias.Set.add n (!cols); expr
+      | _ -> expr) in
+  !cols
+
+let rename_var expr ~replace =
+  traverse expr (fun expr ->
+      match expr with
+      | Var key ->
+        Alias.Map.find ~key replace
+        |> Option.map ~f:var
+        |> Option.value ~default:expr
+      | _ -> expr
+    )
+
+module Constant = struct
+  open Lens_constant
+
+  let bool v = Constant (bool v)
+
+  let int v = Constant (int v)
+
+  let of_value v = Constant (of_value v)
+end
+
+let replace_var expr ~replace =
+  traverse expr (fun expr ->
+      match expr with
+      | Var key ->
+        Alias.Map.find ~key replace
+        |> Option.map ~f:Constant.of_value
+        |> Option.value ~default:expr
+      | _ -> expr
+    )
+
+let rec eval expr get_val =
+  let open Value in
+  match expr with
+  | Constant c -> Lens_constant.to_value c
+  | Var v ->
+    begin
+      try
+        get_val v
+      with NotFound _ -> failwith ("Could not find column " ^ v ^ ".")
+    end
+  | InfixAppl (op, a1, a2) ->
+    let a1 = eval a1 get_val in
+    let a2 = eval a2 get_val in
+    begin
+      match op with
+      | Binary.Equal -> box_bool ( match a1 with `Bool b -> b == unbox_bool a2 | `Int i -> i == unbox_int a2 | `String s -> s = unbox_string a2 | _ -> failwith "Unsupported equality constant.")
```

Comment:
Breaking this into multiple lines, would make it more readable.

Reply:
Changed.